retrofit-mock/src/main/java/retrofit/mock/BehaviorCall.java

```diff
@@ -0,0 +1,183 @@
+
+final class BehaviorCall<T> implements Call<T> {
+  private final Behavior behavior;
+  private final ExecutorService backgroundExecutor;
+  private final Executor callbackExecutor;
+  private final Call<T> delegate;
+
+  private volatile Future<?> task;
+  private volatile boolean canceled;
+  private volatile boolean executed;
+
+  BehaviorCall(Behavior behavior, ExecutorService backgroundExecutor, Executor callbackExecutor,
+      Call<T> delegate) {
+    if (callbackExecutor == null) {
+      callbackExecutor = new Executor() {
+        @Override public void execute(Runnable command) {
+          command.run();
+        }
+      };
+    }
+    this.behavior = behavior;
+    this.backgroundExecutor = backgroundExecutor;
+    this.callbackExecutor = callbackExecutor;
+    this.delegate = delegate;
+  }
+
+  @SuppressWarnings("CloneDoesntCallSuperClone") // We are a final type & this saves clearing state.
+  @Override public Call<T> clone() {
+    return new BehaviorCall<>(behavior, backgroundExecutor, callbackExecutor, delegate.clone());
+  }
+
+  @Override public void enqueue(final Callback<T> callback) {
+    synchronized (this) {
+      if (executed) throw new IllegalStateException("Already executed");
+      executed = true;
+    }
+    task = backgroundExecutor.submit(new Runnable() {
+      @Override public void run() {
+        if (behavior.calculateIsFailure()) {
+          try {
+            Thread.sleep(behavior.calculateDelay(MILLISECONDS));
+          } catch (InterruptedException e) {
+            callbackExecutor.execute(new Runnable() {
+              @Override public void run() {
+                callback.onFailure(new InterruptedIOException("canceled"));
+              }
+            });
+            return;
+          }
+          callbackExecutor.execute(new Runnable() {
+            @Override public void run() {
+              callback.onFailure(behavior.failureException());
+            }
+          });
+        } else {
+          delegate.enqueue(new Callback<T>() {
+            @Override public void onResponse(final Response<T> response) {
+              long sleepMs = behavior.calculateDelay(MILLISECONDS);
+              if (sleepMs > 0) {
+                try {
+                  Thread.sleep(sleepMs);
+                } catch (InterruptedException e) {
+                  callbackExecutor.execute(new Runnable() {
+                    @Override public void run() {
+                      callback.onFailure(new InterruptedIOException("canceled"));
+                    }
+                  });
+                  return;
+                }
+              }
+              callbackExecutor.execute(new Runnable() {
+                @Override public void run() {
+                  callback.onResponse(response);
+                }
+              });
+            }
+
+            @Override public void onFailure(final Throwable t) {
+              long sleepMs = behavior.calculateDelay(MILLISECONDS);
```

Comment:
maybe you want helper methods, `sleepAndCallOnFailure()`, `sleepAndCallOnResponse()` ?